## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [8]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [10]:
# View the column names in the data
print(df.columns)


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [5]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [ ]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# For example, you could look at the data types
# of each column to understand what the data represents.

# View the data types of each column
print(df.dtypes)






In [14]:
# View the number of unique values in each column

print(df.nunique())




first            360
last             477
job              498
phone            999
email            999
client_id        999
order_id       10000
order_date       150
order_week        23
order_year         2
item_id          500
category           5
subcategory       18
unit_price       490
unit_cost        487
unit_weight      393
qty             2911
line_number       10
dtype: int64


In [15]:
# View the number of missing values in each column
print(df.isnull().sum())



first          0
last           0
job            0
phone          0
email          0
client_id      0
order_id       0
order_date     0
order_week     0
order_year     0
item_id        0
category       0
subcategory    0
unit_price     0
unit_cost      0
unit_weight    0
qty            0
line_number    0
dtype: int64


In [17]:
# View the number of rows and columns in the data
print(df.shape)

(54639, 18)


In [27]:
# What three item categories had the most entries?
print(df['category'].value_counts().head(3))



category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64


In [41]:
# For the category with the most entries, which subcategory had the most entries?
# the consumables category has the most entries

# Identify the category with the most entries
most_common_category = df['category'].value_counts().idxmax()

# # For the category with the most entries, find the subcategory with the most entries
most_common_subcategory = df[df['category'] == most_common_category]['subcategory'].value_counts().idxmax()
# Print the results with the number of entries for the category and subcategory
print(f"The category with the most entries is '{most_common_category}' with {df['category'].value_counts().max()} entries.")

# Print the results with the number of entries for the category and subcategory
print(f"The subcategory with the most entries is '{most_common_subcategory}' with {df[df['category'] == most_common_category]['subcategory'].value_counts().max()} entries.")

The category with the most entries is 'consumables' with 23538 entries.
The subcategory with the most entries is 'bathroom supplies' with 6424 entries.


In [49]:
# Which five clients had the most entries in the data?

top_five_clients = df['client_id'].value_counts().head(5)

# Print the results and include the first name, last name and client ID of the clients
print("The five clients with the most entries are:")
for client_id, count in top_five_clients.items():
    client_data = df[df['client_id'] == client_id].iloc[0]
    print(f"Client ID: {client_id}, {client_data['first']} {client_data['last']}, Number of Entries: {count}")


The five clients with the most entries are:
Client ID: 33615, Jessica Reyes, Number of Entries: 220
Client ID: 66037, Angela Everett, Number of Entries: 211
Client ID: 46820, Bryan Myers, Number of Entries: 209
Client ID: 38378, Alexandra Young, Number of Entries: 207
Client ID: 24741, Kendra Garrett, Number of Entries: 207


In [52]:
# Store the client ids of those top 5 clients in a list.
top_five_client_ids = top_five_clients.index.tolist()
print(top_five_client_ids)


[33615, 66037, 46820, 38378, 24741]


In [62]:
# Identify the client with the most entries
client_with_most_entries = df[df['client_id'] == df['client_id'].value_counts().idxmax()]

# Calculate the total units ordered by the client
total_units_ordered = client_with_most_entries['qty'].sum()

# Print the result
print(f"The client with the most entries ordered was client_id\
 {client_with_most_entries.iloc[0]['client_id']},\
 {client_with_most_entries.iloc[0]['first']} {client_with_most_entries.iloc[0]['last']},\
 for a total of {total_units_ordered} units ordered.")



The client with the most entries ordered was client_id 33615, Jessica Reyes, for a total of 64313 units ordered.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [66]:
# Create a column named subtotal that calculates the subtotal for each line using the unit_price and the qty.
# Calculate the subtotal for each line
# Create a new column named 'subtotal' that multiplies the 'unit_price' by the 'qty'
# Display the 'unit_price', 'qty', and 'subtotal' columns

df['subtotal'] = df['unit_price'] * df['qty']
print(df[['unit_price', 'qty', 'subtotal']].head()) 




   unit_price  qty   subtotal
0     1096.80  105  115164.00
1       24.95   21     523.95
2       13.52   39     527.28
3       36.42   29    1056.18
4      195.10   20    3902.00


In [68]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
# Calculate the shipping price for each line and store the results in a new column named 'shipping_price'.
# Display the 'unit_price', 'unit_weight', 'qty', 'total_weight' and 'shipping_price' columns.

# Calculate the total weight for each line
df['total_weight'] = df['unit_weight'] * df['qty']

# Calculate the shipping price based on the total weight
df['shipping_price'] = df['total_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)

# Display the specified columns
print(df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']])


       unit_price  unit_weight   qty  total_weight  shipping_price
0         1096.80         7.50   105        787.50         5512.50
1           24.95         1.49    21         31.29          312.90
2           13.52         1.68    39         65.52          458.64
3           36.42         1.23    29         35.67          356.70
4          195.10        46.43    20        928.60         6500.20
...           ...          ...   ...           ...             ...
54634       83.13         2.25    33         74.25          519.75
54635      206.59        11.70    47        549.90         3849.30
54636       65.66         4.16   475       1976.00        13832.00
54637        1.48        18.04   112       2020.48        14143.36
54638        3.01         2.07  1031       2134.17        14939.19

[54639 rows x 5 columns]


In [70]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
# Calculate the total price for each line
# Create a new column named 'total_price' that adds the 'subtotal' and 'shipping_price' columns and multiplies the result by 1.0925
# Display the 'subtotal', 'shipping_price', and 'total_price' columns

# Calculate the total price for each line
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925

# Display the 'subtotal', 'shipping_price', and 'total_price' columns
print(df[['subtotal', 'shipping_price', 'total_price']].head())

    subtotal  shipping_price    total_price
0  115164.00         5512.50  131839.076250
1     523.95          312.90     914.258625
2     527.28          458.64    1077.117600
3    1056.18          356.70    1543.571400
4    3902.00         6500.20   11364.403500


In [82]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

# Round the 'total_price' column to 2 decimal places
df['total_price'] = df['total_price'].round(2)

# Calculate the total cost for each line
df['total_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

# Display 'unit_cost', 'qty', 'shipping_price', and 'total_cost'
print(df[['unit_cost', 'qty', 'shipping_price', 'total_price', 'total_cost']])



       unit_cost   qty  shipping_price  total_price  total_cost
0         762.71   105         5512.50    131839.08    85597.05
1          15.09    21          312.90       914.26      629.79
2           7.86    39          458.64      1077.12      765.18
3          24.85    29          356.70      1543.57     1077.35
4         108.17    20         6500.20     11364.40     8663.60
...          ...   ...             ...          ...         ...
54634      51.60    33          519.75      3564.87     2222.55
54635     175.46    47         3849.30     14813.24    12095.92
54636      57.31   475        13832.00     49184.90    41054.25
54637       1.22   112        14143.36     15632.71    14280.00
54638       1.61  1031        14939.19     19711.43    16599.10

[54639 rows x 5 columns]


In [85]:
# Create a column for the profit of each line using total_cost and total_price
# Calculate the profit for each line

# Create a new column named 'profit' that subtracts the 'total_cost' from the 'total_price'
df['profit'] = df['total_price'] - df['total_cost']

# Display the 'total_price', 'total_cost', and 'profit' columns
print(df[['order_id', 'total_price', 'total_cost', 'profit']].head())





   order_id  total_price  total_cost    profit
0   8953482    131839.08    85597.05  46242.03
1   8069089       914.26      629.79    284.47
2   1902144      1077.12      765.18    311.94
3   9031802      1543.57     1077.35    466.22
4   1322274     11364.40     8663.60   2700.80


## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [96]:
# Check your work using the totals above
# locate the following order IDs: 2742071, 2173913, 6128929
# display the ['order_id', 'total_price', 'total_cost', 'profit']

order_ids = [2742071, 2173913, 6128929]
filtered_df = df.loc[df['order_id'].isin(order_ids), ['order_id', 'total_price', 'total_cost', 'profit']]

# sum the total_price column, grouping by order_id
total_order_price_subset = filtered_df.groupby('order_id')['total_price'].sum()
total_order_price = df.groupby('order_id')['total_price'].sum()

# display the results
for order_id, total in total_order_price_subset.items():
    print(f"Order {order_id} total: ${total:,.2f}")



Order 2173913 total: $162,388.71
Order 2742071 total: $152,811.89
Order 6128929 total: $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [108]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.


#filtered_df = df.loc[df['client_id'].isin(top_five_client_ids), ['client_id', 'order_id', 'total_price', 'total_cost', 'profit']]
filtered_df = df.loc[df['client_id'].isin(top_five_client_ids)]

total_spent_by_top_five_clients = filtered_df.groupby('client_id')['total_price'].sum()

# display the results
for client_id, total in total_spent_by_top_five_clients.items():
    print(f"Client {client_id} total spent: ${total:,.2f}")



Client 24741 total spent: $82,268,892.02
Client 33615 total spent: $8,377,308.52
Client 38378 total spent: $12,906,550.87
Client 46820 total spent: $9,743,794.36
Client 66037 total spent: $10,259,514.79


In [193]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

# Filter the DataFrame to include only the top 5 clients
top_five_clients_df = df.loc[df['client_id'].isin(top_five_client_ids)]

# Group by client_id and aggregate the required columns
summary_df = top_five_clients_df.groupby('client_id').agg(
    tot_units_purch=('qty', 'sum'),
    tot_shipping_price=('shipping_price', 'sum'),
    tot_rev=('total_price', 'sum'),
    total_cost=('total_cost', 'sum'),
    tot_profit=('profit', 'sum')
).reset_index()

# Display the summary DataFrame
print(summary_df)

   client_id  tot_units_purch  tot_shipping_price      tot_rev   total_cost  \
0      24741           239862          5126448.37  82268892.02  45688899.71   
1      33615            64313          1828984.89   8377308.52   6175313.91   
2      38378            73667          3429455.40  12906550.87   9634720.98   
3      46820            75768          1601448.84   9743794.36   7007191.64   
4      66037            43018          1395151.85  10259514.79   7004482.98   

    tot_profit  
0  36579992.31  
1   2201994.61  
2   3271829.89  
3   2736602.72  
4   3255031.81  


In [195]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['tot_shipping_price', 'tot_rev', 'total_cost', 'tot_profit']

# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    return f"${value/1000000:.6f}M"

# recreate the summary DataFrame for convenience in case I rerun the code
summary_df = top_five_clients_df.groupby('client_id').agg(
    tot_units_purch=('qty', 'sum'),
    tot_shipping_price=('shipping_price', 'sum'),
    tot_rev=('total_price', 'sum'),
    total_cost=('total_cost', 'sum'),
    tot_profit=('profit', 'sum')
).reset_index()

# Apply the currency_format_millions function to only the money columns
summary_df[money_columns] = summary_df[money_columns].map(currency_format_millions)

# Rename the columns to reflect the change in the money format.
summary_df = summary_df.rename(columns={'client_id': 'Client ID', 'tot_units_purch': 'Units', 'tot_shipping_price': 'Shipping (millions)', 'tot_rev': 'Total Revenue (millions)', 'total_cost': 'Total Cost (millions)', 'tot_profit': 'Total Profit (millions)'})

# Display the summary DataFrame to verify the changes
# print(summary_df)

# Rename the columns to reflect the change in the money format. 
# summary_df = top_five_clients_df.groupby('client_id').agg(
#     Units=('qty', 'sum'),
#     Shipping=('shipping_price', 'sum'),
#     tot_rev=('total_price', 'sum'),
#     total_cost=('total_cost', 'sum'),
#     tot_profit=('profit', 'sum')
# ).reset_index().rename(columns={'client_id': 'Client ID', 'Shipping': 'Shipping (millions)', 'tot_rev': 'Total Revenue (millions)', 'total_cost': 'Total Cost (millions)', 'tot_profit': 'Total Profit (millions)'})

# Display the summary DataFrame
print(summary_df)




   Client ID   Units Shipping (millions) Total Revenue (millions)  \
0      24741  239862          $5.126448M              $82.268892M   
1      33615   64313          $1.828985M               $8.377309M   
2      38378   73667          $3.429455M              $12.906551M   
3      46820   75768          $1.601449M               $9.743794M   
4      66037   43018          $1.395152M              $10.259515M   

  Total Cost (millions) Total Profit (millions)  
0           $45.688900M             $36.579992M  
1            $6.175314M              $2.201995M  
2            $9.634721M              $3.271830M  
3            $7.007192M              $2.736603M  
4            $7.004483M              $3.255032M  


In [198]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_summary_df = summary_df.sort_values('Total Profit (millions)', ascending=False)
print(sorted_summary_df)


   Client ID   Units Shipping (millions) Total Revenue (millions)  \
0      24741  239862          $5.126448M              $82.268892M   
2      38378   73667          $3.429455M              $12.906551M   
4      66037   43018          $1.395152M              $10.259515M   
3      46820   75768          $1.601449M               $9.743794M   
1      33615   64313          $1.828985M               $8.377309M   

  Total Cost (millions) Total Profit (millions)  
0           $45.688900M             $36.579992M  
2            $9.634721M              $3.271830M  
4            $7.004483M              $3.255032M  
3            $7.007192M              $2.736603M  
1            $6.175314M              $2.201995M  
